src/createExplorerBase.ts

```diff
@@ -0,0 +1,146 @@
+import path from 'path';
+import { loaders } from './loaders';
+import { getPropertyByPath } from './getPropertyByPath';
+import {
+  CosmiconfigResult,
+  ExplorerOptions,
+  ExplorerOptionsSync,
+  Cache,
+  LoadedFileContent,
+} from './types';
+import { Loader } from './index';
+
+class ExplorerBase<T extends ExplorerOptions | ExplorerOptionsSync> {
```

Comment:
Curious: why is this `T extends A | B` rather than `A | B`?

Reply:
Generics are used here because `ExplorerOptions` are not compatible with `ExplorerOptionsSync` so you have to explicitly tell typescript what object it is expecting so it can return the correct type. At least that is how I understand it.

If you try removing the generic and replacing it with `ExplorerOptions | ExplorerOptionsSync` You'll see several errors in `createExplorerSync.ts`.